src/query/expression/src/kernels/filter.rs

```diff
@@ -424,4 +423,215 @@ impl Column {
 
         StringColumn::new(data.into(), offsets.into())
     }
+
+    fn copy_continuous_bits(
+        ptr: &mut *mut u8,
+        src: &[u8],
+        mut dst_idx: usize,
+        mut src_idx: usize,
+        len: usize,
+    ) -> (u8, usize) {
+        let mut unset_bits = 0;
+        let chunks = BitChunks::new(src, src_idx, len);
+        chunks.iter().for_each(|chunk| {
+            unset_bits += chunk.count_zeros();
+            unsafe {
+                copy_advance_aligned(&chunk as *const _ as *const u8, ptr, 8);
+            }
+        });
+
+        let mut remainder = chunks.remainder_len();
+        dst_idx += len - remainder;
+        src_idx += len - remainder;
+
+        let mut buf = 0;
+        unsafe {
+            while remainder > 0 {
+                if (*src.as_ptr().add(src_idx >> 3) & BIT_MASK[src_idx & 7]) != 0 {
+                    buf |= BIT_MASK[dst_idx % 8];
+                } else {
+                    unset_bits += 1;
+                }
+                src_idx += 1;
+                dst_idx += 1;
+                remainder -= 1;
+                if dst_idx % 8 == 0 {
+                    store_advance_aligned(buf, ptr);
+                    buf = 0;
+                }
+            }
+        }
+        (buf, unset_bits as usize)
+    }
+
+    // low-level API using unsafe to improve performance.
+    fn filter_boolean_types(bitmap: &Bitmap, filter: &Bitmap) -> Bitmap {
+        debug_assert_eq!(bitmap.len(), filter.len());
+        let num_rows = filter.len() - filter.unset_bits();
+        if num_rows == bitmap.len() {
+            return bitmap.clone();
+        } else if num_rows == 0 {
+            return Bitmap::new();
+        }
+        // Fast path.
+        if num_rows <= bitmap.len()
+            && (bitmap.unset_bits() == 0 || bitmap.unset_bits() == bitmap.len())
+        {
+            let mut bitmap = bitmap.clone();
+            bitmap.slice(0, num_rows);
+            return bitmap;
+        }
+
+        let capacity = num_rows.saturating_add(7) / 8;
+        let mut builder: Vec<u8> = Vec::with_capacity(capacity);
+        let mut builder_ptr = builder.as_mut_ptr();
+        let mut builder_idx = 0;
+        let mut unset_bits = 0;
+        let mut buf = 0;
+
+        let (bitmap_slice, bitmap_offset, _) = bitmap.as_slice();
+        let mut bitmap_idx = 0;
+
+        let (mut filter_slice, filter_offset, mut filter_length) = filter.as_slice();
+        unsafe {
+            if filter_offset > 0 {
```

Comment:
need a test case to cover this branch, using `slice` to create bitmap with non-zero offset

Reply:
Done.